Garbage Collector

MrMinemeet opened the pull request:
Implements Mark & Sweep based on Deutsch-Schorr-Waite algorithm.

MrMinemeet committed:
Basic structure for Garbage Collection

MrMinemeet committed:
Fix SegFault caused by non-dynamic casting

With dynamic casting, it works. Otherwise C++ would regularly throw SegFaults

MrMinemeet committed:
Add basic GC call test

MrMinemeet committed:
Start implementation of GC

MrMinemeet committed:
Add SlideBlock as actual class for testing

MrMinemeet committed:
Add mark(Block* rootpointer)

Ass the mark-method based on 02. Garbage Collection (Page 22).

The implementation uses more than the three local variables stated in the slides, but I found it a bit more read- and understandable this way.

MrMinemeet committed:
Add "isMarked" info to dump

MrMinemeet committed:
Add Sweep phase

Work-in-Progress implementation based on 02. Garbage Collection (Page 27)

MrMinemeet committed:
Add sweep

Work in Progress.

Also had do adjust some other stuff as there were some things wrong

MrMinemeet committed:
Adjusted preprocessor ifs

Also removed the "debug" filling of blocks. That just breaks stuff.

Also enabled debug mode again

BioGustav commented on src/Block.cpp:
```diff
@@ -32,19 +33,79 @@ std::string Block::ToString() const {
 bool Block::isFreeBlock() const {
 	FreeBlock* fb = (FreeBlock*) this;
-	return (void*) fb->typeDescriptor == fb->dataPosition();
+	FreeBlock fbA = *fb;
+	TypeDescriptor* td = fb->getTypeDescriptor();
+	void* dataPart = fb->getDataPart();
```
those lines seem to be not needed (?)

MrMinemeet replied:
yea, correct. They were from debugging

BioGustav commented on src/Block.cpp:
```diff
@@ -32,19 +33,79 @@ std::string Block::ToString() const {
 void* Block::getDataPart() const {
 	// "this + sizeof(Block)" is the effective getDataPart for the block
-	return (void*) (this + headerSize());
+	return (void*) ((char*) this + headerSize());
```
ah yes.. pointercasts

MrMinemeet replied:
This cast is needed, otherwise it is basically `this + headerSize() * sizeof(Block)`.
Meaning, the pointer would be too far off. With the `(char*)` cast it is `headerSize() * sizeof(char*))` which results in just `headerSize()`

BioGustav commented on src/Block.cpp:
```diff
@@ -32,19 +33,79 @@ std::string Block::ToString() const {
+TypeDescriptor* Block::getTypeDescriptor() const {
+	// Remove the LSB from the rawTypeDescriptor pointer
+	// TODO: Do this in a nice way
+	return (TypeDescriptor*) (((uintptr_t) rawTypeDescriptor >> 1) << 1);
```
for now ok, I mean if it works it works

BioGustav replied:
maybe AND it with a flipped 1? eg.: 0xFE for 1 byte

edit: just read that we could use ~1 (not 1)

BioGustav commented on src/Block.cpp:
```diff
@@ -32,19 +33,79 @@ std::string Block::ToString() const {
+ * Marks the block by setting the LSB of the type descriptor to 1
+ */
+void Block::mark(bool b) {
+	// Set the LSB to 1
```
this comment may is a bit confusing here: one may think this sets LSB always to 1 (which would only require the first branch of the if)

MrMinemeet replied:
Adjusted the comment

BioGustav commented on src/Block.cpp:
```diff
@@ -32,19 +33,79 @@ std::string Block::ToString() const {
+	if (b)
+		rawTypeDescriptor = (TypeDescriptor*) ((uintptr_t) rawTypeDescriptor | 1);
+	else
+		rawTypeDescriptor = (TypeDescriptor*) (((uintptr_t) rawTypeDescriptor >> 1) << 1);
```
Just flip the bit if its state is not as desired.
```suggestion
	if (this->isMarked() != b)
		rawTypeDescriptor = (TypeDescriptor*) ((uintptr_t) rawTypeDescriptor ^ 1);
```

BioGustav commented on src/Heap.cpp:
```diff
@@ -297,4 +306,166 @@ std::string Heap::getTypeDescriptorName(TypeDescriptor* typeDescriptor) const {
+		return;
+	}
+
+	/* Instead of marking the currently visited pointer inside "Block" I decided to use a map here.
+	 * Since this is only needed for the marking phase, I think it is okay to use a map here.
+	 */
```
But you are marking anyways?

BioGustav commented on src/main.cpp:
```diff
@@ -10,43 +11,61 @@ int main() {
-	std::cout << heap.ToString() << std::endl;
+	UsedBlock* b1 = heap.alloc("SlideBlock");
+	// I just use "new" here to trigger the constructor
+	// FIXME: Something is wrong when the SlideBlock constructor is called and sets the pointers to null.
```
You mean the pointers inside slideblock?

MrMinemeet replied:
Just ignore the `FIXME`. They were my helping comments and I forgot to remove them

MrMinemeet committed:
Add Suggestions

Also add '#include <cstdint>' because it needs this on some systems (at least on the one of @BioGustav)
